### PR TITLE
refactor: use explicit checks in ica controller upgrade handlers

### PR DESCRIPTION
**Pull request by colin-axner**
## Description

ref: #5462 

### Commit Message / Changelog Entry

```text
type: commit message
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/apps/27-interchain-accounts/controller/keeper/handshake.go [hunk 1]
```diff
@@ -162,30 +187,49 @@ func (k Keeper) OnChanUpgradeInit(ctx sdk.Context, portID, channelID string, ord
...
+// The following may be changed:
+// - tx type (must be supported)
+// - encoding (must be supported)
+//
+// The following may not be changed:
+// - controller connectionID
+// - host connectionID
+// - interchain account address
+// - interchain account version
 func (k Keeper) OnChanUpgradeAck(ctx sdk.Context, portID, channelID, counterpartyVersion string) error {
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/controller/keeper/handshake.go:227]: didn't get a chance to update the upgrade ack tests yet

File: modules/apps/27-interchain-accounts/controller/keeper/handshake.go [hunk 2]
```diff
@@ -162,30 +187,49 @@ func (k Keeper) OnChanUpgradeInit(ctx sdk.Context, portID, channelID string, ord
 		return "", err
 	}
 
-	if err := icatypes.ValidateControllerMetadata(ctx, k.channelKeeper, connectionHops, metadata); err != nil {
-		return "", errorsmod.Wrap(err, "invalid metadata")
+	// ValidateControllerMetadata will ensure the interchain account version has not changed and that the
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/controller/keeper/handshake.go:190]: we could pull here down into a helper func `ValidateUpgradeMetadata`, pulling out the relevant checks from `ValidateControllerMetadata`, but left as duplication between upgrade init and upgrade ack for now

File: modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go [hunk 3]
```diff
@@ -481,9 +481,19 @@ func (suite *KeeperTestSuite) TestOnChanUpgradeInit() {
 	var (
 		path     *ibctesting.Path
 		metadata icatypes.Metadata
+		version  string
 		order    channeltypes.Order
 	)
 
+	// updateMetadata is a helper function which modifies the metadata stored in the channel version
+	// and marshals it into a string to pass to OnChanUpgradeInit as the counterpartyVersion string.
+	updateMetadata := func(modificationFn func(*icatypes.Metadata)) {
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/controller/keeper/handshake_test.go:495]: updated to match upgrade ack

**Review by chatton — APPROVED** [review]
LGTM, very nice test coverage :heart:

**Review by crodriguezvega — APPROVED** [review]
Nice work, @colin-axner

File: modules/apps/27-interchain-accounts/controller/keeper/handshake.go [hunk 4]
```diff
@@ -162,30 +187,49 @@ func (k Keeper) OnChanUpgradeInit(ctx sdk.Context, portID, channelID string, ord
...
 
-	if currentMetadata.ControllerConnectionId != connectionHops[0] {
-		return "", errorsmod.Wrap(connectiontypes.ErrInvalidConnectionIdentifier, "proposed connection hop must not change")
+	if currentMetadata.ControllerConnectionId != proposedMetadata.ControllerConnectionId {
+		return "", errorsmod.Wrap(connectiontypes.ErrInvalidConnection, "proposed controller connection ID must not change")
+	}
+
+	if currentMetadata.HostConnectionId != proposedMetadata.HostConnectionId {
+		return "", errorsmod.Wrap(connectiontypes.ErrInvalidConnection, "proposed host connection ID must not change")
 	}
```
Comment (crodriguezvega) [inline modules/apps/27-interchain-accounts/controller/keeper/handshake.go:208]: I don't feel too strongly about it, but if these checks are not reachable in either `OnChanUpgradeInit` and `OnChanUpgradeAck`, couldn't we just remove them and add a comment above the call to `ValidateControllerMetadata` to explain that this function protects against these changes?

**Pull request merged**